### ISSUE TITLE
Question about image warping and blending weight

### ISSUE THREAD

**Issue by kwea123** [zhengqili/Neural-Scene-Flow-Fields#19]
Hi, the idea of warping and having a blending weight predicting how to merge static and dynamic results is a good idea, however I have been confused by some operations in the implementation. Initially I was not sure if these confusing parts lead to bad results, until now I see several issues concerning the performance on own datasets, so I decided to post this issue, hoping to provide some insights that could potentially solve some of the issues.

1. Predicting the blending weight by the static network seems really counterintuitive since dynamic objects move inside the scene independently of the static network. A more reasonable way is to predict this weight by the **dynamic** network as suggested by [this paper](https://arxiv.org/pdf/2105.06468.pdf). Or a even simpler way is to totally remove this blending weight and use the addition strategy as in NeRF-W. I did a short experiment on the difference between these two blending strategies (blending weight vs addition), and find that addition produces better reconstruction and novel view results. Although this finding might not be always correct for different data, at least I think predicting the blending weight by the static network is not the ideal way to go.
2. For image warping, you do blending in the current timestamp
https://github.com/zhengqili/Neural-Scene-Flow-Fields/blob/4cb2ef4314ca4f569d7d72f1b4ddc3e756dc3815/nsff_exp/render_utils.py#L1019-L1022
but only render the dynamic parts of t-1 and t+1
https://github.com/zhengqili/Neural-Scene-Flow-Fields/blob/4cb2ef4314ca4f569d7d72f1b4ddc3e756dc3815/nsff_exp/render_utils.py#L1055-L1057
this is really confusing, basically it means there are two rendering pipelines, and the network should learn to maximize the performance of both pipelines. It is hard to tell how this causes problem exactly, but I doubt that this reduces the final performance that uses blending. In my opinion, suppose that having the static network gives better result, you should use blending in current time rendering **and in image warping too**. Again, I don't know whether changing this yields better result, but this part is confusing.

I would like to know @zhengqili 's opinion on these points, and maybe suggest the users to try these modifications to see if it solves some problem.

**Comment by zhengqili** [comment]
Hi , Thanks for the suggestions!

(1) In terms of blending weight from static or dynamic model: I actually tried both setups (In fact, initially I designed predicting blending weight from the dynamic model). In my opinion, in order to make time-variant blending weight working, in the final combined rendering loss, we have to use pairwise photometric loss again, otherwise, supervising blending weight of time t with only frame observation at t is not sufficient. Instead, time-invariant blending weight can lead to simpler final combined rendering loss (only per-frame RGB loss without pairwise photometric losses), which makes losses, in general, less confusing and cluttered. Furthermore, what time-invariant weight works here is that for the entire motion trajectory volume, the weight for the dynamic model should be close to 1, and this will not cause problem during view synthesis, since for empty space, even if we use dynamic model, its predicted density is still, letting us see the background.

(2) In terms of image warping, the basic idea is that we train a dynamic model using depth, flow, and photometric consistency to learn the RGB/density of entire images (i.e., the rendering from function "raw2outputs_warp"), this pipeline can be independent of static model and final combined rendering. The final rendering from function "raw2outputs_blending" is to integrate blending weight to "ideally" mask out the opacity/color from dynamic nerf and instead use the color/opacity learned from a static model. Without blending weight, either dynamic model will dominate everywhere or there will be duplications or color-mix for static regions since we will render color/opacity from both networks at the same time. 

(3) I am very interested in how to remove blending weight because I believe a more principle and CORRECT way is similar to NeRF-W. However, I spent months trying to remove this weight but never works well for most videos I have. If we remove weights, we need to move all the depth, flow, and pairwise photometric consistency to the combined rendering (mentioned in (2) above, I put these losses on rendering from the dynamic model only, which is independent of the combined rendering), but if I do so, the rendering quality of moving regions and fg/bg separation can be much worse. I am very curious how to make it work, or did I miss something here?

**Comment by kwea123** [comment]
For points (1) and (2), I think I understand what you mean, let me try to conceive a concrete scenario where it could cause problem.

For (3), my implementation is pretty simple, remove the blending weight and every other code remains exactly the same. Like you said, yes, we need to move all the depth, flow, and pairwise photometric consistency to the combined rendering. In my experiments that I mentioned above
> I did a short experiment on the difference between these two blending strategies (blending weight vs addition), and find that addition produces better reconstruction and novel view results

it yields better result, but I cannot show you here because the data is internal. I will find some time to replicate it on a public data!

**Comment by zhengqili** [comment]
Very interesting! Have you tried the addition approach on the kid-running example I provide, does it work as well?

I did try the addition approach a long time ago, so I am thinking about what part can make our implementation different. The only one I remember that can cause different results is computing 2D optical flow projected from 3D points. I am curious how you implement this function "compute_optical_flow", or shall I add/remove some regularization term?

**Comment by kwea123** [comment]
yeah, I will try on kid-running first and let you know the result!

**Comment by kwea123** [comment]
First of all I tried removing the blending weight here:
https://github.com/zhengqili/Neural-Scene-Flow-Fields/blob/7df65546bda2d48b443e5441b972eb96fe989a38/nsff_exp/render_utils.py#L808-L809
and train the network. As you said the result presents a lot of artifacts as can be seen in the training plots.

Then I suspect that it is because of the warping method, that is, if I remove the blending weight, then in warping I must compose both static and dynamic using the same adding strategy like in NeRF-W.
However, I find it hard to correctly implement this based on your code because there are too many methods that I need to modify, thus I'm afraid of missing some and the result will be meaningless. Therefore I implemented this based on my nerf_pl repo.
So basically the idea is like my questions above, I only removed the blending weight, and use adding strategy to compose static and dynamic, and other structures/losses should be roughly the same as what you have here. Here's the result on kid-running after 150k steps (8hr on 1 gpu by the way):

https://user-images.githubusercontent.com/11364490/120870102-fb9b6000-c5d2-11eb-854f-2e97fc700f0d.mp4

There is some ghosting artifact remaining, I'm trying to figure out the reason, but generally the quality is not too bad imo. Also the fg/bg separation is good like the image below.
![a](https://user-images.githubusercontent.com/11364490/120870212-4fa64480-c5d3-11eb-86ea-55c8baed1724.png)

I wonder if you have encountered this kind of layered ghosting issue, which will help me debugging.

**Comment by kwea123** [comment]
After adding the entropy regularization as you suggested in #18, the ghosting disappears, but now the shadow and the bottom part of the feet gets segmented into the static part, so it still looks weird.

https://user-images.githubusercontent.com/11364490/120945769-3e953900-c775-11eb-8410-dc0e096ed135.mp4

However I believe this should not be far from the final result I will get (will keep investigating).
I observe several advantages over the current method:
1. The reconstruction has better quality in terms of PSNR (the training time is still 8h on 1 gpu, or 150k steps of batchsize 512).
<p align="center">
  <img src="https://user-images.githubusercontent.com/11364490/120945372-c5491680-c773-11eb-82d7-29bdf8a7a74e.png", width="32%">
  <img src="https://user-images.githubusercontent.com/11364490/120945610-adbe5d80-c774-11eb-8555-818d79291737.png", width="32%">
  <img src="https://user-images.githubusercontent.com/11364490/120945350-ac406580-c773-11eb-8d71-1c568796adcd.jpg", width="32%">
  <br>
  <sup>Left: GT. Center: my method (PSNR=33.86). Right: pretrained model of your method (PSNR=30.42).</sup>
</p>
2. The static part is totally static and does not change (even a little) over time, but with your method the static structure gradually deforms as the sequence moves on (look at the left side of the car for example). Drag the progress bar quickly, the difference will be more obvious.

https://user-images.githubusercontent.com/11364490/120945871-96cc3b00-c775-11eb-9ab8-2019fed0d379.mp4

I will open my code once the remaining artifacts disappear. I think my conclusion from the results until now is that **removing the blending weight and use static+dynamic compositing in the warping operation too** leads to better results. I have some background experiments that I have done on **multiview** data where this strategy exhibits the best result. Now to make this work on monocular data, I think other regularizations are needed, but the best rendering process should still be the adding strategy without blending weight.

Now please consider this issue being solved. I will post a last comment when I fix everything and release my code.
Please do not hesitate if you have further comments/suggestions, thank you!

**Comment by kwea123** [comment]
I just had another **important** findings. Empirically I find that the flows are sometimes attributed to faraway points (points whose z is large), this is counter-intuitive and should not happen in real world, where we expect faraway points to move much slower than near points. The small flow regularization of the current implementation doesn't do the trick because even a little flow in NDC space of far points translates to large displacement in Euclidean space, and this causes the network to stuck in that bad minima. Therefore, I **explicitly zero the flow if the z of a point is greater than 0.95** (altenately, maybe we can encourage the network to *learn* small flow in far region by weighting far points more), and now the results are even better than what I showed above.

Finally, the code will be opened [here](https://github.com/kwea123/nsff_pl) in a few days.

**Comment by kwea123** [comment]
The first version is out, I will keep improving and trying on other datasets (such as the ones in the issues)

**Comment by zhengqili** [comment]
Thanks a lot! This is amazing! I will post a link to your repo so that more people can see it :)
